shared-bindings/traceback/__init__.c

```diff
@@ -0,0 +1,170 @@
+//|
+//| """
+//| ...
+//|
+
+//| def print_exception(etype: Type[BaseException], value: BaseException, tb: TracebackType,
+//|                     limit: Optional[int] = None, file: Optional[io.FileIO] = None, chain: Optional[bool] = True) -> None:
+//|
+//|     """Prints exception information and stack trace entries.
+//|
+//|     .. note: Setting `chain` will have no effect as chained exceptions are not yet implemented.
+//|
+//|     :param Type[BaseException] etype: This is ignored and inferred from the type of ``value``.
+//|     :param BaseException value: The exception. Must be an instance of `BaseException`.
+//|     :param TracebackType tb: The traceback object. If `None`, the traceback will not be printed.
+//|     :param int limit: Print up to limit stack trace entries (starting from the caller’s frame) if limit is positive.
+//|                       Otherwise, print the last ``abs(limit)`` entries. If limit is omitted or None, all entries are printed.
+//|     :param io.FileIO file: If file is omitted or `None`, the output goes to `sys.stderr`; otherwise it should be an open
+//|                            file or file-like object to receive the output.
+//|     :param bool chain: If `True` then chained exceptions will be printed.
+//|
+//|     """
+//|     ...
+//|
+STATIC mp_obj_t traceback_print_exception(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_etype, ARG_value, ARG_tb, ARG_limit, ARG_file, ARG_chain };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_etype, MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_value, MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_tb,    MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_limit, MP_ARG_OBJ,  {.u_obj = mp_const_none} },
+        { MP_QSTR_file,  MP_ARG_OBJ,  {.u_obj = mp_const_none} },
+        { MP_QSTR_chain, MP_ARG_BOOL, {.u_bool = true} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    mp_obj_t exc = args[ARG_value].u_obj;
+    if (!mp_obj_is_exception_instance(exc)) {
+        mp_raise_TypeError(translate("invalid exception"));
+    }
+
+    mp_print_t print = mp_plat_print;
+    if (args[ARG_file].u_obj != mp_const_none) {
+        #if MICROPY_PY_IO && MICROPY_PY_SYS_STDFILES
+        mp_get_stream_raise(args[ARG_file].u_obj, MP_STREAM_OP_WRITE);
+        print.data = MP_OBJ_TO_PTR(args[ARG_file].u_obj);
+        print.print_strn = mp_stream_write_adaptor;
+        #else
+        mp_raise_NotImplementedError(translate("file write is not available"));
+        #endif
+    }
+
+    mp_int_t limit = 0;
+    bool print_tb = true;
+    if (args[ARG_limit].u_obj != mp_const_none) {
+        if (!mp_obj_get_int_maybe(args[ARG_limit].u_obj, &limit)) {
+            mp_raise_TypeError(translate("limit should be an int"));
+        }
+        print_tb = !(limit == 0);
+    }
+
```

Comment:
Please put everything below in shared-module. It's easier to read shared bindings if it only does argument validation.